include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1583 @@
+        // this scary warning:
+        //
+        // "
+        // ADOL-C Warning: Tapeless: Setting numDir could change memory
+        // allocation of derivatives in existing adoubles and may lead to
+        // erroneous results or memory corruption
+        // "
+        //
+        // So we use this dummy function to configure this setting before
+        // we create and initialize our class data
+        const std::size_t n_live_variables = adtl::refcounter::getNumLiveVar();
+        if (n_live_variables == 0)
+          {
+            adtl::setNumDir(n_directional_derivatives);
+          }
+        else
+          {
+            // So there are some live active variables floating around. Here we
+            // check if we ask to increase the number of number of computable
+            // directional derivatives. If this really is necessary then its
+            // absolutely vital that there exist no live variables before doing
+            // so.
+            const std::size_t n_set_directional_derivatives = adtl::getNumDir();
+            if (n_directional_derivatives > n_set_directional_derivatives)
+              AssertThrow(
+                n_live_variables == 0,
+                ExcMessage(
+                  "There are currently " +
+                  Utilities::to_string(n_live_variables) +
+                  " live "
+                  "adtl::adouble variables in existence. They currently "
+                  "assume " +
+                  Utilities::to_string(n_set_directional_derivatives) +
+                  " directional derivatives "
+                  "but you wish to increase this to " +
+                  Utilities::to_string(n_directional_derivatives) +
+                  ". \n"
+                  "To safely change (or more specifically in this case, "
+                  "increase) the number of directional derivatives, there "
+                  "must be no tapeless doubles in local/global scope."));
+          }
+#    else
+        // If ADOL-C is not configured with tapeless number reference counting
+        // then there is no way that we can guarentee that the following call
+        // is safe. No comment... :-/
+        adtl::setNumDir(n_directional_derivatives);
+#    endif
+      }
+
+#  else
+
+      template <typename ADNumberType>
+      static typename std::enable_if<ADNumberTraits<ADNumberType>::type_code ==
+                                     NumberTypes::adolc_tapeless>::type
+      configure_tapeless_mode(const unsigned int n_directional_derivatives)
+      {
+        AssertThrow(false, ExcRequiresAdolC());
+      }
+
+#  endif
+
+    } // namespace internal
+
+
+    // Specialization for auto-differentiable numbers that use
+    // reverse mode to compute the first derivatives (and, if supported,
+    // forward mode for the second).
+    template <typename ADNumberType, typename ScalarType>
+    struct TapelessDrivers<
```

Comment:
doxygen style for class docs, `@copydoc` for member docs in the same way as above

Reply:
This is all encapsulated in a `#ifndef DOXYGEN`, so we don't build documentation for the class specialisations.

Reply:
(That's [this line](https://github.com/dealii/dealii/pull/6976/files#diff-42fec071a85b1a667a206111406371e6R490))